### ISSUE TITLE
GCC Flag: -Wshadow

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/picongpu#1042]
After #1039, running or adding [-Wshadow](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html) to the code might be a good idea:

```
-Wshadow
    Warn whenever a local variable or type declaration shadows another variable, parameter, type, class member (in C++), or instance variable (in Objective-C) or whenever a built-in function is shadowed. Note that in C++, the compiler warns if a local variable shadows an explicit typedef, but not if it shadows a struct/class/enum. 
```

**Comment by BenjaminW3** [comment]
:+1:
Other compilers (e.g. MSVC 2015) do this by default and spam the build output :smile:

**Comment by ax3l** [comment]
hm, why did I expect that this would happen :smiley:

**Comment by Flamefire** [comment]
Tried -Wshadow and it DOES spam the build output with hundreds of messages...

**Comment by ax3l** [comment]
all right, can you pls investigate the urgent ones?
PRs are very much welcome (besides adding the flag to `CMakeLists.txt` itself as soon as it throws)! :)

**Comment by Flamefire** [comment]
I don't think this is really useful.  I guess a lot of messages come from constructors like:
Foo(int x): x(x)...

Those are ok (IMO) unless one wants to enforce a naming scheme like "x_" for member variables.

**Comment by ax3l** [comment]
... that's while doing `: x(x)` is bad style, but no-one believes me ^^

I read that these positions should throw warnings, too. Can you do some research if one can suppress those for now and check for in function/member `{...}` warnings?

**Comment by Flamefire** [comment]
A quick research did not show anything besides telling that it is bad practice. So one can only rename the members or the parameters from foo to foo_ style or make up a better convention.

I'll try to find something else but if there isn't a solution is required. Wshadow is a good warning which should be used IMO.

**Comment by ax3l** [comment]
I think we should try to keep our code free from Wshadow and use member replacements.

Libraries with similar complexity (and massive C++ meta programming) such as [kokkos](https://github.com/kokkos/kokkos) even get even more crazy flags done, too:
```
Primary tested compiler are passing in release mode
with warnings as errors. We are using the following set
of flags:
GCC:   -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits
       -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized
Intel: -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
Clang: -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized
```

**Comment by Flamefire** [comment]
I just checked 2 random files from Kokkos (one of them was the random implementation...) and it uses underscores (frankly sometimes for parameters, sometimes for members)
So if nothing uses appended underscores yet, I'd silence those warnings by changing the parameter names. This is less intrusive

**Comment by psychocoderHPC** [comment]
IMO we should mark member variables with `m_` like it is done in `alpaka` https://github.com/ComputationalRadiationPhysics/alpaka/blob/master/include/alpaka/mem/buf/BufCudaRt.hpp#L117-L120

**Comment by psychocoderHPC** [comment]
Please keep in mind that underscore at the end of the name is reserved in `PMacc`for a instance of an identifier. e.g. particle[position_],...

**Comment by Flamefire** [comment]
Right. But one use of underscore is "const int x_ = gBlocks.x() - 2 * base.getGuardingSuperCells();"
I also think that marking all members with m_ is not that good. Think of e.g.

    struct Point{
       int m_x, m_y, m_z;
    };

    //Usage
    pt.m_x = 5; //Ugly!

For private members this would be ok though, also not nice.

And it also requires some serious effort with unknown implications. Refactoring tools are almost useless due to templating so one might miss instances that compile anyway (e.g. initialization, shadowed members,...)

**Comment by ax3l** [comment]
what about a *prefix* `_` for member vars?

**Comment by Flamefire** [comment]
As mentioned, it would still look ugly on access. Normally any prefix should not be required (--> Smurf naming?) Setters would use something like "setX(int newX){ x=newX;}" So any access would still look like "pt.x = 5" which is much better than "pt._x = 5" isn't it?
So I'd still rather change the parameter names leaving members as good looking as they can get. Only exception are private members as those are not seen from the outside. And there still is the issue, that changing member names is an error-prone, repetitive, and non-automateable process while changing parameters is at least a bit better (refactoring tools can be used, very limited scope)

One comment from http://stackoverflow.com/questions/2476381/c-constructor-initialization-list-strangeness: "if your code's readability depends on knowing from its name alone whether a variable is a member or not, I'd say it's time to clean up and refactor."

**Comment by psychocoderHPC** [comment]
@ax3l prefix `_` is not allowed this results in compiler bugs in the past. I will add a link to the old pull request tomorrow.

**Comment by ax3l** [comment]
@psychocoderHPC true, thx!
actually the argument to better change `params` is quite reasonable.

**Comment by ax3l** [comment]
an other argument be @psychocoderHPC (also concerning my argument "rather use getter/setter"):

the problem with renaming parameters is, that those are really visible to users of the class/function. members on the other hand are in libPMacc *barely* used directly from other classes/externaly, we are encapsulating most of those as one should do.

**Comment by Flamefire** [comment]
Short evaluation of the number of files with Wshadow warnings:

    grep -Eo ".*?/src/.*?\..pp" output.txt | sort -u | wc -l

shows 121 files.

The only safe way to fix that is by renaming the constructor parameters and/or checking for any other errors. If there are none then one could rename the members and make it compile afterwards. Will be quite some work...

**Comment by ax3l** [comment]
in `picongpu`, the main affected files are in `fields` and `plugins`. `libPMacc` has about the same amount of files (equally distributed). a few more in `mallocMC` (fixed in https://github.com/ComputationalRadiationPhysics/mallocMC/pull/95) and in `libSplash` (https://github.com/ComputationalRadiationPhysics/libSplash/issues/190) and three in the cuda RT headers for 6.5 ;)

I will fix the `picongpu` `field` files now with an `m_` prefix.

**Comment by ax3l** [comment]
all affected files: ```grep -Re "[[:blank:]]\(\w\+\)([[:blank:]]*\1[[:blank:]]*)" src/```

**Comment by ax3l** [comment]
still more warnings need to be fixed.
